db/de.go

```diff
@@ -105,7 +105,28 @@ func (d *DEDatabase) AddUserDataUsageBatch(context context.Context, usages map[s
 		placeholders = append(placeholders, "(?::text, ?::bigint)")
 		startargs = append(startargs, usr, usg)
 	}
-	startcte := "WITH new_usages (username, usage) AS (VALUES " + strings.Join(placeholders, ",") + ")"
+	nonzero_usages := "new_nonzero_usages (username, usage) AS (VALUES " + strings.Join(placeholders, ",") + ")"
+
+	// Add a 0 for any user whose most recent total is > 0 but who doesn't appear in the batch we got from the ICAT
+	new_usages2, nuargs, err := psql.Select().
+		Column("us.username").
+		Column("?", 0).
+		From(d.Table("user_data_usage", "udu")).
+		Join(fmt.Sprintf("%s ON (us.id = udu.user_id)", d.Table("users", "us"))).
+		LeftJoin("new_nonzero_usages ON users.username = new_nonzero_usages.username").
+		Where(squirrel.Gt{"total": 0}).
+		Where("us.username BETWEEN ? AND ?", start, end).
```

Comment:
I was having trouble figuring out exactly how Squirrel would generate this query, so I copied the code to a test program and ran it with some log messages. I ended up getting these messages:

```
2022/11/02 16:04:21 Query: WITH new_nonzero_usages (username, usage) AS (VALUES ($1::text, $2::bigint),($3::text, $4::bigint)), new_usages (username, usage) AS (SELECT username, usage from new_nonzero_usages UNION ALL SELECT us.username, $1 FROM public.user_data_usage AS udu JOIN public.users AS us ON (us.id = udu.user_id) LEFT JOIN new_nonzero_usages ON us.username = new_nonzero_usages.username WHERE total > $2 AND us.username BETWEEN $3 AND $4 AND time = (SELECT MAX(time) FROM user_data_usage u2 WHERE u2.user_id = udu.user_id)) INSERT INTO public.user_data_usage AS d (total,time,user_id) SELECT new_usages.usage AS total, $5 AT TIME ZONE (SELECT current_setting('TIMEZONE')) AS time, u.id FROM public.users AS u JOIN new_usages ON (new_usages.username = u.username) RETURNING d.id, d.total, d.user_id, (SELECT username from users WHERE id = d.user_id) as username, d.time AT TIME ZONE (SELECT current_setting('TIMEZONE')) AS time, d.last_modified AT TIME ZONE (SELECT current_setting('TIMEZONE')) AS last_modified
2022/11/02 16:04:21 Args: [user1@iplantcollaborative.org 401398 user2@iplantcollaborative.org 0 0 0 user1@iplantcollaborative.org user2@iplantcollaborative.org 2022-11-02 16:04:21.97566 -0700 MST m=+0.150461626]
2022/11/02 16:04:22 unable to get the user data usages: Error running query: pq: operator does not exist: character varying <= bigint
```

I used actual usernames in the test program, but they've been redacted here for privacy.

Looking at the query, it appears that the only operator in the query that would cause a `<=` comparison would be `BETWEEN`, which is being used to compare the usernames. The argument indexes in this case are `$3` and `$4`. It appears that the third and fourth arguments in the argument list are `user2@iplantcollaborative.org` and `0`, so the zero must be causing the error.

These aren't the intended arguments for that clause, so it appears that something is throwing off the argument indices, although I'm not entirely sure how to fix it.

Admittedly, I could have messed something up when I was copying the code. 😆 I'll be happy to share the test program that I used with you if you'd like to look it over.

Reply:
This comment in an issue for Squirrel sounds like it might work for this: https://github.com/Masterminds/squirrel/issues/21#issuecomment-791539537

Reply:
Cool, the solution described in the comment that I linked got past that error.

Reply:
Now I'm encountering this error:

Reply:
Ah, you're right that it shouldn't be using `psql` for this subquery, but just `squirrel`. I think your second error is because your test data for `new_nonzero_usages` has a 0 value in it, so it's probably duplicating that value. Not sure though.

Reply:
Ah, yes, it does have a zero in it, so that probably is the problem. I'll update the test data and try again.

Reply:
I updated the test data, and I'm still getting the error.

Reply:
oh, I totally see what's happening. in the midst of everything else I forgot to actually exclude anything that had data already in `new_nonzero_usages` :laughing: 

I'm making that and the squirrel issue and then I'll run some tests and push the changes up. sorry about the confusion here

Reply:
Your latest commit resolves this problem. 👍